sentence.go

```diff
@@ -74,69 +79,137 @@ func (s BaseSentence) TalkerID() string {
 	if startIndex != 0 {
-		return BaseSentence{}, fmt.Errorf("nmea: sentence does not start with a '$' or '!'")
+		return BaseSentence{}, errors.New("nmea: sentence does not start with a '$' or '!'")
 	}
-	sumSepIndex := strings.Index(raw, ChecksumSep)
-	if sumSepIndex == -1 {
-		return BaseSentence{}, fmt.Errorf("nmea: sentence does not contain checksum separator")
+	checksumSepIndex := strings.Index(raw, ChecksumSep)
+	rawFields := raw[startIndex+1:]
+	checksumRaw := ""
+	if checksumSepIndex != -1 {
+		rawFields = raw[startIndex+1 : checksumSepIndex]
+		checksumRaw = strings.ToUpper(raw[checksumSepIndex+1:])
 	}
+	// TODO: fields can contain encoded chars that need to be unescaped. `^` 0x5e is escape character for HEX representation of ISO/IEC 8859-1 (ASCII) characters.
+	// TODO: `^` itself is escaped as `^5e` and `,` is escaped as `^2c` etc. All reserved characters should be escaped/unescaped (See wikipedia https://en.wikipedia.org/wiki/NMEA_0183#Message_structure)
+	fields := strings.Split(rawFields, FieldSep)
+
 	var (
-		fieldsRaw   = raw[startIndex+1 : sumSepIndex]
-		fields      = strings.Split(fieldsRaw, FieldSep)
-		checksumRaw = strings.ToUpper(raw[sumSepIndex+1:])
-		checksum    = Checksum(fieldsRaw)
+		talkerID string
+		typ      string
 	)
-	// Validate the checksum
-	if checksum != checksumRaw {
-		return BaseSentence{}, fmt.Errorf(
-			"nmea: sentence checksum mismatch [%s != %s]", checksum, checksumRaw)
+	if p.ParsePrefix == nil {
+		talkerID, typ, err = DefaultParsePrefix(fields[0])
+	} else {
+		talkerID, typ, err = p.ParsePrefix(fields[0])
 	}
-	talker, typ := parsePrefix(fields[0])
-	return BaseSentence{
-		Talker:   talker,
+	if err != nil {
+		return BaseSentence{}, err
+	}
+	sentence := BaseSentence{
+		Talker:   talkerID,
 		Type:     typ,
 		Fields:   fields[1:],
 		Checksum: checksumRaw,
 		Raw:      raw,
 		TagBlock: tagBlock,
-	}, nil
+	}
+	if p.CheckCRC == nil {
+		err = DefaultCheckCRC(sentence, rawFields)
+	} else {
+		err = p.CheckCRC(sentence, rawFields)
+	}
+	if err != nil {
+		return BaseSentence{}, err
+	}
+	return sentence, nil
+}
+
+// DefaultCheckCRC is default implementation for checking sentence Checksum
+func DefaultCheckCRC(sentence BaseSentence, rawFields string) error {
+	if sentence.Checksum == "" {
+		return fmt.Errorf("nmea: sentence does not contain checksum separator")
+	}
+	if checksum := Checksum(rawFields); checksum != sentence.Checksum {
+		return fmt.Errorf("nmea: sentence checksum mismatch [%s != %s]", checksum, sentence.Checksum)
+	}
+	return nil
 }
 
-// parsePrefix takes the first field and splits it into a talker id and data type.
-func parsePrefix(s string) (string, string) {
-	if strings.HasPrefix(s, "PMTK") {
-		return "PMTK", s[4:]
+// DefaultParsePrefix takes in the sentence first field (NMEA0183 Address) and splits it into a talker id and sentence type.
+func DefaultParsePrefix(prefix string) (string, string, error) {
```

Comment:
The `Default` prefix is unnecessary. This function stands on its own outside of the `SentenceParser` implementation. The fact that this is used as the default implementation for `SentenceParser.ParsePrefix` should be mentioned on as a doc on the field, not encoded into the function name.

Reply:
renamed